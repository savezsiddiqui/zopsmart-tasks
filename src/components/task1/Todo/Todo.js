import Card from '../TodoCard/ToDoCard'
import React, { Component } from 'react'
import TodoStyles from '../Todo.module.css'

class Todo extends Component {
  constructor(props) {
    super(props)
    this.state = {
      todo: [],
      todoInput: ''
    }
  }

  componentDidMount() {
    const dataString = localStorage.getItem('todos')
    if (dataString !== null) {
      const todos = JSON.parse(dataString)
      this.setState({ todo: todos })
    }
  }

  _onChange = (e) => {
    this.setState({
      todoInput: e.target.value
    })
  }

  _onKeyDown = (e) => {
    if (e.key === 'Enter' && this.state.todoInput.length > 0) {
      const newTodos = [
        {
          text: this.state.todoInput,
          checked: false,
          id: new Date()
        },
        ...this.state.todo
      ]
      localStorage.setItem('todos', JSON.stringify(newTodos))
      this.setState({
        todo: newTodos,
        todoInput: ''
      })
    }
  }

  deleteTodo = (id) => {
    const newTodos = this.state.todo.filter((todo) => todo.id !== id)
    localStorage.setItem('todos', JSON.stringify(newTodos))

    this.setState({
      todo: newTodos
    })
  }

  editTodo = (text, id) => {
    const newTodos = this.state.todo.map((todo) =>
      todo.id === id ? { ...todo, text } : todo
    )
    localStorage.setItem('todos', JSON.stringify(newTodos))
    this.setState({
      todo: newTodos
    })
  }

  checkTodo = (id) => {
    const newTodos = this.state.todo.map((todo) =>
      todo.id === id
        ? {
            ...todo,
            checked: !todo.checked
          }
        : todo
    )

    localStorage.setItem('todos', JSON.stringify(newTodos))

    this.setState({
      todo: newTodos
    })
  }

  render() {
    const cardList = this.state.todo.map(({ text, checked, id }) => (
      <Card
        text={text}
        checked={checked}
        key={id}
        checkFn={() => this.checkTodo(id)}
        delFn={() => this.deleteTodo(id)}
        editFn={(text) => this.editTodo(text, id)}
      />
    ))

    return (
      <div className={TodoStyles.root}>
        <section className={TodoStyles['fixed-top']}>
          <h1 className={TodoStyles['display-4']}>TODO</h1>
          <input
            type="text"
            id="todo-input"
            className={TodoStyles['custom-input']}
            placeholder="add todo"
            onKeyDown={this._onKeyDown}
            onChange={this._onChange}
            value={this.state.todoInput}
          />
        </section>
        <section className={TodoStyles['todo-container']}>{cardList}</section>
      </div>
    )
  }
}

export default Todo
