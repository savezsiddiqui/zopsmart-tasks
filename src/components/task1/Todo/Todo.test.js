import { fireEvent, render } from '@testing-library/react'
import renderer from 'react-test-renderer'
import Todo from './Todo'
import React from 'react'

test('should render App correctly', () => {
  const tree = renderer.create(<Todo />).toJSON()
  expect(tree).toMatchSnapshot()
})

test('should add todo', () => {
  const { getByPlaceholderText, getByText } = render(<Todo />)
  const input = getByPlaceholderText('add todo')
  fireEvent.change(input, { target: { value: 'test-todo-1' } })
  fireEvent.keyDown(input, { key: 'Enter', code: 'Enter' })
  expect(getByText('test-todo-1')).toBeInTheDocument()
})

test('should not add todo', () => {
  const tree = renderer.create(<Todo />).toJSON()
  const { getByPlaceholderText } = render(<Todo />)
  const input = getByPlaceholderText('add todo')
  fireEvent.change(input, { target: { value: '' } })
  fireEvent.keyDown(input, { key: 'Enter', code: 'Enter' })
  expect(tree).toMatchSnapshot()
})

test('should check todo', () => {
  const { getByPlaceholderText, getByText, getByTestId } = render(<Todo />)
  const input = getByPlaceholderText('add todo')
  fireEvent.change(input, { target: { value: 'test-todo-2' } })
  fireEvent.keyDown(input, { key: 'Enter', code: 'Enter' })
  fireEvent.click(getByText('test-todo-2'))
  expect(getByTestId('strike-through')).toBeInTheDocument()
})

test('should delete todo', () => {
  const { getByPlaceholderText, getByText, getAllByTestId } = render(<Todo />)
  const input = getByPlaceholderText('add todo')
  fireEvent.change(input, { target: { value: 'test-todo-3' } })
  fireEvent.keyDown(input, { key: 'Enter', code: 'Enter' })
  fireEvent.click(getAllByTestId('delete-button')[0])
  expect(() => getByText('test-todo-3')).toThrowError()
})

test('should edit todo', () => {
  const { getByPlaceholderText, getByText, getAllByTestId } = render(<Todo />)
  const input = getByPlaceholderText('add todo')
  fireEvent.change(input, { target: { value: 'test-todo-4' } })
  fireEvent.keyDown(input, { key: 'Enter', code: 'Enter' })
  fireEvent.click(getAllByTestId('edit-button')[0])
  const editInput = getAllByTestId('edit-input')[0]
  fireEvent.change(editInput, { target: { value: 'test-todo-4-edited' } })
  fireEvent.keyDown(editInput, { key: 'Enter', code: 'Enter' })
  expect(getByText('test-todo-4-edited')).toBeInTheDocument()
})

test('should not add edited todo', () => {
  const tree = renderer.create(<Todo />).toJSON()
  const { getByPlaceholderText, getAllByTestId } = render(<Todo />)
  const input = getByPlaceholderText('add todo')
  fireEvent.change(input, { target: { value: 'test-todo-5' } })
  fireEvent.keyDown(input, { key: 'Enter', code: 'Enter' })
  fireEvent.click(getAllByTestId('edit-button')[0])
  const editInput = getAllByTestId('edit-input')[0]
  fireEvent.change(editInput, { target: { value: '' } })
  fireEvent.keyDown(editInput, { key: 'Enter', code: 'Enter' })
  expect(tree).toMatchSnapshot()
})
