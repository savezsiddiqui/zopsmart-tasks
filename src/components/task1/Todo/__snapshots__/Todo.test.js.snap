// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not add edited todo 1`] = `
<div
  className="root"
>
  <section
    className="fixed-top"
  >
    <h1
      className="display-4"
    >
      TODO
    </h1>
    <input
      className="custom-input"
      id="todo-input"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="add todo"
      type="text"
      value=""
    />
  </section>
  <section
    className="todo-container"
  >
    <div
      className="card"
    >
      <div
        className="todo-wrapper"
      >
        <div
          className="todo"
          onClick={[Function]}
        >
          <i
            className="far far fa-circle"
          />
          <p>
            test-todo-4-edited
          </p>
        </div>
        <div
          className="options"
        >
          <i
            className="far far fa-trash-alt fa-trash-alt"
            data-testid="delete-button"
            onClick={[Function]}
          />
          <i
            className="fas fas fa-pencil-alt fa-pencil-alt"
            data-testid="edit-button"
            onClick={[Function]}
          />
        </div>
      </div>
    </div>
    <div
      className="card"
    >
      <div
        className="todo-wrapper"
      >
        <div
          className="todo"
          onClick={[Function]}
        >
          <i
            className="far far fa-check-circle"
          />
          <p>
            <strike
              data-testid="strike-through"
            >
              test-todo-2
            </strike>
          </p>
        </div>
        <div
          className="options"
        >
          <i
            className="far far fa-trash-alt fa-trash-alt"
            data-testid="delete-button"
            onClick={[Function]}
          />
          <i
            className="fas fas fa-pencil-alt fa-pencil-alt"
            data-testid="edit-button"
            onClick={[Function]}
          />
        </div>
      </div>
    </div>
    <div
      className="card"
    >
      <div
        className="todo-wrapper"
      >
        <div
          className="todo"
          onClick={[Function]}
        >
          <i
            className="far far fa-circle"
          />
          <p>
            test-todo-1
          </p>
        </div>
        <div
          className="options"
        >
          <i
            className="far far fa-trash-alt fa-trash-alt"
            data-testid="delete-button"
            onClick={[Function]}
          />
          <i
            className="fas fas fa-pencil-alt fa-pencil-alt"
            data-testid="edit-button"
            onClick={[Function]}
          />
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`should not add todo 1`] = `
<div
  className="root"
>
  <section
    className="fixed-top"
  >
    <h1
      className="display-4"
    >
      TODO
    </h1>
    <input
      className="custom-input"
      id="todo-input"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="add todo"
      type="text"
      value=""
    />
  </section>
  <section
    className="todo-container"
  >
    <div
      className="card"
    >
      <div
        className="todo-wrapper"
      >
        <div
          className="todo"
          onClick={[Function]}
        >
          <i
            className="far far fa-circle"
          />
          <p>
            test-todo-1
          </p>
        </div>
        <div
          className="options"
        >
          <i
            className="far far fa-trash-alt fa-trash-alt"
            data-testid="delete-button"
            onClick={[Function]}
          />
          <i
            className="fas fas fa-pencil-alt fa-pencil-alt"
            data-testid="edit-button"
            onClick={[Function]}
          />
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`should render App correctly 1`] = `
<div
  className="root"
>
  <section
    className="fixed-top"
  >
    <h1
      className="display-4"
    >
      TODO
    </h1>
    <input
      className="custom-input"
      id="todo-input"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="add todo"
      type="text"
      value=""
    />
  </section>
  <section
    className="todo-container"
  />
</div>
`;
